modules/puppetca_token_whitelisting/puppetca_token_whitelisting_autosigner.rb

```diff
@@ -0,0 +1,123 @@
+require 'jwt'
+require 'openssl'
+
+module ::Proxy::PuppetCa::TokenWhitelisting
+  class Autosigner
+    include ::Proxy::Log
+    include ::Proxy::Util
+
+    JWT_ALGORITHM = 'RS512'
+
+    def tokens_file
+      Proxy::PuppetCa::TokenWhitelisting::Plugin.settings.tokens_file
+    end
+
+    def sign_all
+      Proxy::PuppetCa::TokenWhitelisting::Plugin.settings.sign_all
+    end
+
+    def smartproxy_cert
+      @certificate ||= OpenSSL::PKey::RSA.new File.read cert_file
+    end
+
+    def storage
+      Proxy::PuppetCa::TokenWhitelisting::TokenStorage.new tokens_file
+    end
+
+    def token_ttl
+      Proxy::PuppetCa::TokenWhitelisting::Plugin.settings.token_ttl
+    end
+
+    def cert_file
+      return Proxy::SETTINGS.ssl_private_key.to_s if Proxy::PuppetCa::TokenWhitelisting::Plugin.settings.certificate.nil?
+
+      file = Proxy::PuppetCa::TokenWhitelisting::Plugin.settings.certificate
+      unless File.exist?(file)
+        File.write file, OpenSSL::PKey::RSA.generate(2048)
+      end
+      file
+    end
+
+    # Invalidate a token based on the certname
+    def disable certname
+      storage.lock do
+        tokens = storage.read
+        tokens.delete_if do |token|
+          begin
+            decoded = JWT.decode(token, smartproxy_cert.public_key, true, algorithm: JWT_ALGORITHM)
+            decoded.first['certname'] == certname
+          rescue JWT::ExpiredSignature
+            true
+          end
+        end
+        storage.unsave_write tokens
```

Comment:
Any reason this should be unsafe?

Reply:
The whole block is already locked so nothing changes the file between read and write. You would not be able to safely write here, as that would try to lock the file again which is already locked.

Reply:
I wonder if the storage should have an operation where you pass in a block to delete tokens. It feels a bit like you're doing storage operations in the signer here.

Reply:
Added a `TokenStorage::remove_if` method, take a look.